mtags/src/main/scala-3/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -0,0 +1,418 @@
+   * @param indexedContext
+   * @param config
+   * @param parent the parent tree node of the pattern match, for example `Apply(_, _)` when in
+   *               `List(1).foreach { cas@@ }`, used as fallback to compute the type of the selector when
+   *               it's `EmptyTree`.
+   * @param patternOnly `true` if we want completions without `case` keyword
+   * @param hasBind `true` when `case _: @@ =>`, if hasBind we don't need unapply completions
+   */
+  def contribute(
+      selector: Tree,
+      completionPos: CompletionPos,
+      typedTree: Tree,
+      indexedContext: IndexedContext,
+      config: PresentationCompilerConfig,
+      parent: Tree,
+      patternOnly: Boolean = false,
+      hasBind: Boolean = false,
+      // doFilterText: Boolean = false,
+  ): List[CompletionValue] =
+    import indexedContext.ctx
+    val pos = completionPos.sourcePos
+    val definitions = indexedContext.ctx.definitions
+    val text = pos.source.content().mkString
+    val clientSupportsSnippets = config.isCompletionSnippetsEnabled()
+    lazy val autoImportsGen = AutoImports.generator(
+      pos,
+      text,
+      typedTree,
+      indexedContext,
+      config,
+    )
+    val completionGenerator = CompletionValueGenerator(
+      indexedContext,
+      completionPos,
+      clientSupportsSnippets,
+      patternOnly,
+      hasBind,
+    )
+    val parents: Parents = selector match
+      case EmptyTree =>
+        val seenFromType = parent match
+          case TreeApply(fun, _) if fun.tpe != null && !fun.tpe.isErroneous =>
+            fun.tpe
+          case _ =>
+            parent.seenFrom(parent.symbol)._1
+        seenFromType.paramInfoss match
+          case (head :: Nil) :: _
+              if definitions.isFunctionType(head) || head.isRef(
+                definitions.PartialFunctionClass
+              ) =>
+            val dealiased = head.widenDealias
+            val argTypes =
+              head.argTypes.init
+            new Parents(argTypes, definitions)
+          case _ =>
+            new Parents(NoType, definitions)
+      case sel => new Parents(sel.tpe, definitions)
+
+    val result = ListBuffer.empty[CompletionValue]
+    val isVisited = mutable.Set.empty[Symbol]
+
+    def visit(sym: Symbol, name: String, autoImports: List[l.TextEdit]): Unit =
+      val isValid = !isVisited(sym) && !parents.isParent(sym)
+        && (sym.is(Case) || sym.is(Flags.Module) || sym.isClass)
+        && parents.isSubClass(sym, false)
+      def recordVisit(s: Symbol): Unit =
+        if s != NoSymbol && !isVisited(s) then
+          isVisited += s
+          recordVisit(s.moduleClass)
+          recordVisit(s.sourceModule)
+      if isValid then
+        recordVisit(sym)
+        result += completionGenerator.toCompletionValue(
+          sym,
+          name,
+          autoImports,
+        )
+      end if
+    end visit
+    val selectorSym = parents.selector.typeSymbol
+
+    // Step 1: walk through scope members.
```

Comment:
Why we're walking through scope members? I thought it doesn't related to case completions?

Reply:
For things like in this test 
https://github.com/scalameta/metals/blob/0443f440379712314c85a2b77a6707177cb326a0/tests/cross/src/test/scala/tests/pc/CompletionCaseSuite.scala#L42-L69

But if we remove walking through scope members, it would enable `enum` completions without the need of any additional changes (they are seen as sealedStringDescendants, but are not a subclass co `visit` filters them out). The only issue then is to change the way we name patterns (remove unnecessary things like `scala.util.None` instead of `None`)

Reply:
Ah, so this is for non-sealed classes, right?

- First and foremost, I don't think we don't need to provide an exhaustive match for non-sealed classes like this.
  - Because there's nothing like "exhaustiveness" in non-sealed classes and traits.
- Does this work if the class definitions are different files? (I don't think so)
  - I think `children` would provide the direct subclasses (maybe?)

So my suggestion is, to remove this walking through scope symbols for non-sealed types, and let's enable to complete `enum`s :+1:

Reply:
Oops, nevermind, this is for case completion, not for match exhaustive one :bow: